tests/e2e/test_read.py

```diff
@@ -39,29 +44,87 @@ def container_client(account_url):
-    return f"{account_url}/{container_client.container_name}/{blob_client.blob_name}"
+@pytest.fixture(scope="module")
+def large_blob(account_url, container_client):
+    return upload_blob(
+        account_url, container_client, sample_data(_PARTITIONED_DOWNLOAD_THRESHOLD * 2)
+    )
+
+
+@pytest.fixture(scope="module")
+def small_with_newlines_blob(account_url, container_client):
+    return upload_blob(account_url, container_client, sample_data_with_newlines(20, 2))
 
 
 @pytest.fixture
-def blob_io(blob_url):
-    yield BlobIO(blob_url=blob_url, mode="rb")
+def blob(request):
+    return request.getfixturevalue(f"{request.param}_blob")
 
 
 def random_resource_name(name_length=8):
-    return "".join(random.choices(string.ascii_lowercase + string.digits, k=name_length))
+    return "".join(
+        random.choices(string.ascii_lowercase + string.digits, k=name_length)
+    )
+
+
+def sample_data(data_length=20):
+    return os.urandom(data_length)
+
+
+def sample_data_with_newlines(data_length=20, num_lines=1):
+    lines = []
+    for i in range(num_lines):
+        lines.append(sample_data(int(data_length / num_lines)))
+    return b"\n".join(lines)
+
+
+def upload_blob(account_url, container_client, data):
+    blob_name = random_resource_name()
+    blob_client = container_client.get_blob_client(blob=blob_name)
+    blob_client.upload_blob(data)
+    url = f"{account_url}/{container_client.container_name}/{blob_name}"
+
+    return Blob(data=data, url=url)
 
 
 class TestRead:
-    def test_reads_all_data(self, blob_io, sample_data):
-        with blob_io as f:
-            assert f.read() == sample_data
+    @pytest.mark.parametrize(
+        "blob",
+        [
+            "small",
+            "large",
+        ],
+        indirect=True,
+    )
+    def test_reads_all_data(self, blob):
+        with BlobIO(blob.url, "rb") as f:
+            assert f.read() == blob.data
+            assert f.tell() == len(blob.data)
+
+    @pytest.mark.parametrize("n", [1, 5, 20, 21])
+    def test_read_n_bytes(self, small_blob, n):
+        with BlobIO(small_blob.url, "rb") as f:
+            for i in range(0, len(small_blob.data), n):
+                assert f.read(n) == small_blob.data[i : i + n]
+                if n < len(small_blob.data):
+                    assert f.tell() == i + n
+
+    @pytest.mark.parametrize("n", [1, 5, 20, 21])
+    def test_random_seeks_and_reads(self, small_blob, n):
+        with BlobIO(small_blob.url, "rb") as f:
+            f.seek(n)
+            assert f.read() == small_blob.data[n:]
+            if n < len(small_blob.data):
+                assert f.tell() == len(small_blob.data)
+
+    def test_read_using_iter(self, small_with_newlines_blob):
+        with BlobIO(small_with_newlines_blob.url, "rb") as f:
+            data = f.read().splitlines()
+            expected_data = small_with_newlines_blob.data.splitlines()
+            for i in data:
+                assert data == expected_data
```

Comment:
For the iter test, let's update it to this:
```python
    def test_read_using_iter(self, small_with_newlines_blob):
        with BlobIO(small_with_newlines_blob.url, "rb") as f:
            lines = [line for line in f]
            expected_lines = small_with_newlines_blob.data.splitlines(keepends=True)
            assert lines == expected_lines
```
The main things are that:
1. We still want to be using the iter on the `BlobIO` object since that is custom to this library
2. This one is more subtle but it is a stronger assertion to assert the entire list instead of doing it one at a time because the `data` could be a subset of `expected_data` and still pass the test even though it is not the same content